<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />



<title>Getting Started</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Getting Started</h1>



<p><code>rsynthbio</code> is an R package that provides a convenient
interface to the <a href="https://www.synthesize.bio/">Synthesize
Bio</a> API, allowing users to generate realistic gene expression data
based on specified biological conditions. This package enables
researchers to easily access AI-generated transcriptomic data for
various modalities including bulk RNA-seq and single-cell RNA-seq.</p>
<p>Alternatively, you can AI generate datasets from our <a href="https://app.synthesize.bio/datasets/">web platform</a>.</p>
<div id="how-to-install" class="section level2">
<h2>How to install</h2>
<p>You can install <code>rsynthbio</code> from CRAN:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">&quot;rsynthbio&quot;</span>)</span></code></pre></div>
<p>If you want the development version, you can install using the
<code>remotes</code> package to install from GitHub:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span>(<span class="st">&quot;remotes&quot;</span> <span class="sc">%in%</span> <span class="fu">installed.packages</span>())) {</span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a>  <span class="fu">install.packages</span>(<span class="st">&quot;remotes&quot;</span>)</span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a>}</span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a>remotes<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">&quot;synthesizebio/rsynthbio&quot;</span>)</span></code></pre></div>
<p>Once installed, load the package:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="fu">library</span>(rsynthbio)</span></code></pre></div>
</div>
<div id="authentication" class="section level2">
<h2>Authentication</h2>
<p>Before using the Synthesize Bio API, you need to set up your API
token. The package provides a secure way to handle authentication:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="co"># Securely prompt for and store your API token</span></span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a><span class="co"># The token will not be visible in the console</span></span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a><span class="fu">set_synthesize_token</span>()</span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a></span>
<span id="cb4-5"><a href="#cb4-5" tabindex="-1"></a><span class="co"># You can also store the token in your system keyring for persistence</span></span>
<span id="cb4-6"><a href="#cb4-6" tabindex="-1"></a><span class="co"># across R sessions (requires the &#39;keyring&#39; package)</span></span>
<span id="cb4-7"><a href="#cb4-7" tabindex="-1"></a><span class="fu">set_synthesize_token</span>(<span class="at">use_keyring =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<p>Loading your API key for a session.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="co"># In future sessions, load the stored token</span></span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a><span class="fu">load_synthesize_token_from_keyring</span>()</span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a></span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a><span class="co"># Check if a token is already set</span></span>
<span id="cb5-5"><a href="#cb5-5" tabindex="-1"></a><span class="fu">has_synthesize_token</span>()</span></code></pre></div>
<p>You can obtain an API token by registering at <a href="https://app.synthesize.bio">Synthesize Bio</a>.</p>
<div id="security-best-practices" class="section level3">
<h3>Security Best Practices</h3>
<p>For security reasons, remember to clear your token when you’re
done:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="co"># Clear token from current session</span></span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a><span class="fu">clear_synthesize_token</span>()</span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a></span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a><span class="co"># Clear token from both session and keyring</span></span>
<span id="cb6-5"><a href="#cb6-5" tabindex="-1"></a><span class="fu">clear_synthesize_token</span>(<span class="at">remove_from_keyring =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<p>Never hard-code your token in scripts that will be shared or
committed to version control.</p>
</div>
</div>
<div id="designing-queries-for-models" class="section level2">
<h2>Designing Queries for Models</h2>
<div id="choosing-a-modality" class="section level3">
<h3>Choosing a Modality</h3>
<p>The modality (data type to generate) is specified in the query using
<code>get_valid_query()</code>:</p>
<ul>
<li><strong><code>bulk</code></strong>: Bulk RNA-seq (asynchronous under
the hood, returned as data frames)</li>
<li><strong><code>single-cell</code></strong>: Single-cell RNA-seq
(asynchronous under the hood, returned as data frames)</li>
</ul>
<p>You can check which modalities are available programmatically:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="co"># Check available modalities</span></span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a><span class="fu">get_valid_modalities</span>()</span></code></pre></div>
<p>You do not need to specify any internal API slugs. The library maps
modalities to the appropriate model endpoints automatically.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="co"># Create a bulk query</span></span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a>bulk_query <span class="ot">&lt;-</span> <span class="fu">get_valid_query</span>(<span class="at">modality =</span> <span class="st">&quot;bulk&quot;</span>)</span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a>bulk <span class="ot">&lt;-</span> <span class="fu">predict_query</span>(bulk_query, <span class="at">as_counts =</span> <span class="cn">TRUE</span>)</span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a></span>
<span id="cb8-5"><a href="#cb8-5" tabindex="-1"></a><span class="co"># Create a single-cell query</span></span>
<span id="cb8-6"><a href="#cb8-6" tabindex="-1"></a>sc_query <span class="ot">&lt;-</span> <span class="fu">get_valid_query</span>(<span class="at">modality =</span> <span class="st">&quot;single-cell&quot;</span>)</span>
<span id="cb8-7"><a href="#cb8-7" tabindex="-1"></a>sc <span class="ot">&lt;-</span> <span class="fu">predict_query</span>(sc_query, <span class="at">as_counts =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
</div>
<div id="creating-a-query" class="section level3">
<h3>Creating a Query</h3>
<p>The structure of the query required by the API is fixed for the
currently supported model. You can use <code>get_valid_query()</code> to
get a correctly structured example list.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="co"># Get the example query structure</span></span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a>example_query <span class="ot">&lt;-</span> <span class="fu">get_valid_query</span>()</span>
<span id="cb9-3"><a href="#cb9-3" tabindex="-1"></a></span>
<span id="cb9-4"><a href="#cb9-4" tabindex="-1"></a><span class="co"># Inspect the query structure</span></span>
<span id="cb9-5"><a href="#cb9-5" tabindex="-1"></a><span class="fu">str</span>(example_query)</span></code></pre></div>
<p>The query consists of:</p>
<ol style="list-style-type: decimal">
<li><strong><code>modality</code></strong>: The type of gene expression
data to generate (“bulk” or “single-cell”)</li>
<li><strong><code>mode</code></strong>: The prediction mode that
controls how expression data is generated:
<ul>
<li><strong>“sample generation”</strong>: Generates realistic-looking
synthetic data with measurement error (bulk only)</li>
<li><strong>“mean estimation”</strong>: Provides stable mean estimates
of expression levels (bulk and single-cell)</li>
</ul></li>
<li><strong><code>inputs</code></strong>: A list of biological
conditions to generate data for</li>
</ol>
<p>Each input contains <code>metadata</code> (describing the biological
sample) and <code>num_samples</code> (how many samples to generate).</p>
<blockquote>
<p>See the <a href="#query-parameters">Query Parameters</a> section
below for detailed documentation on <code>mode</code> and other optional
query fields.</p>
</blockquote>
</div>
<div id="making-a-prediction" class="section level3">
<h3>Making a Prediction</h3>
<p>Once your query is ready, you can send it to the API to generate gene
expression data:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a>result <span class="ot">&lt;-</span> <span class="fu">predict_query</span>(query, <span class="at">as_counts =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<p>This result will be a list of two dataframes: <code>metadata</code>
and <code>expression</code></p>
</div>
<div id="understanding-the-async-api" class="section level3">
<h3>Understanding the Async API</h3>
<p>Behind the scenes, the API uses an <strong>asynchronous
model</strong> to handle queries efficiently:</p>
<ol style="list-style-type: decimal">
<li>Your query is submitted to the API, which returns a query ID</li>
<li>The function automatically polls the status endpoint (default: every
2 seconds)</li>
<li>When the query completes, results are downloaded from a signed
URL</li>
<li>Data is parsed and returned as R data frames</li>
</ol>
<p>All of this happens automatically when you call
<code>predict_query()</code>.</p>
</div>
<div id="controlling-async-behavior" class="section level3">
<h3>Controlling Async Behavior</h3>
<p>You can customize the polling behavior if needed:</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="co"># Increase timeout for large queries (default: 900 seconds = 15 minutes)</span></span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a>result <span class="ot">&lt;-</span> <span class="fu">predict_query</span>(</span>
<span id="cb11-3"><a href="#cb11-3" tabindex="-1"></a>  query,</span>
<span id="cb11-4"><a href="#cb11-4" tabindex="-1"></a>  <span class="at">poll_timeout_seconds =</span> <span class="dv">1800</span>, <span class="co"># 30 minutes</span></span>
<span id="cb11-5"><a href="#cb11-5" tabindex="-1"></a>  <span class="at">poll_interval_seconds =</span> <span class="dv">5</span> <span class="co"># Check every 5 seconds instead of 2</span></span>
<span id="cb11-6"><a href="#cb11-6" tabindex="-1"></a>)</span></code></pre></div>
</div>
<div id="valid-metadata-keys" class="section level3">
<h3>Valid Metadata Keys</h3>
<p>The input metadata is a list of lists. This is the full list of valid
metadata keys:</p>
<p><strong>Biological:</strong></p>
<ul>
<li><code>age_years</code></li>
<li><code>cell_line_ontology_id</code></li>
<li><code>cell_type_ontology_id</code></li>
<li><code>developmental_stage</code></li>
<li><code>disease_ontology_id</code></li>
<li><code>ethnicity</code></li>
<li><code>genotype</code></li>
<li><code>race</code></li>
<li><code>sample_type</code> (“cell line”, “organoid”, “other”, “primary
cells”, “primary tissue”, “xenograft”)</li>
<li><code>sex</code> (“male”, “female”)</li>
<li><code>tissue_ontology_id</code></li>
</ul>
<p><strong>Perturbational:</strong></p>
<ul>
<li><code>perturbation_dose</code></li>
<li><code>perturbation_ontology_id</code></li>
<li><code>perturbation_time</code></li>
<li><code>perturbation_type</code>
(“coculture”,“compound”,“control”,“crispr”,“genetic”,“infection”,“other”,“overexpression”,“peptide
or biologic”,“shrna”,“sirna”)</li>
</ul>
<p><strong>Technical:</strong></p>
<ul>
<li><code>study</code> (Bioproject ID)</li>
<li><code>library_selection</code> (e.g., “cDNA”, “polyA”, “Oligo-dT” -
see <a href="https://ena-docs.readthedocs.io/en/latest/submit/reads/webin-cli.html#permitted-values-for-library-selection" class="uri">https://ena-docs.readthedocs.io/en/latest/submit/reads/webin-cli.html#permitted-values-for-library-selection</a>)</li>
<li><code>library_layout</code> (“PAIRED”, “SINGLE”)</li>
<li><code>platform</code> (“illumina”)</li>
</ul>
</div>
<div id="valid-metadata-values" class="section level3">
<h3>Valid Metadata Values</h3>
<p>The following are the valid values or expected formats for selected
metadata keys:</p>
<table>
<colgroup>
<col width="11%" />
<col width="88%" />
</colgroup>
<thead>
<tr class="header">
<th>Metadata Field</th>
<th>Requirement / Example</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>cell_line_ontology_id</code></td>
<td>Requires a <a href="https://www.cellosaurus.org/">Cellosaurus
ID</a>.</td>
</tr>
<tr class="even">
<td><code>cell_type_ontology_id</code></td>
<td>Requires a <a href="https://www.ebi.ac.uk/ols4/ontologies/cl">CL
ID</a>.</td>
</tr>
<tr class="odd">
<td><code>disease_ontology_id</code></td>
<td>Requires a <a href="https://www.ebi.ac.uk/ols4/ontologies/mondo">MONDO ID</a>.</td>
</tr>
<tr class="even">
<td><code>perturbation_ontology_id</code></td>
<td>Must be a valid Ensembl gene ID (e.g.,
<code>ENSG00000156127</code>), <a href="https://www.ebi.ac.uk/chebi/">ChEBI ID</a> (e.g.,
<code>CHEBI:16681</code>), <a href="https://www.ebi.ac.uk/chembl/">ChEMBL ID</a> (e.g.,
<code>CHEMBL1234567</code>), or <a href="https://www.ncbi.nlm.nih.gov/taxonomy">NCBI Taxonomy ID</a> (e.g.,
<code>9606</code>).</td>
</tr>
<tr class="odd">
<td><code>tissue_ontology_id</code></td>
<td>Requires a <a href="https://www.ebi.ac.uk/ols4/ontologies/uberon">UBERON ID</a>.</td>
</tr>
</tbody>
</table>
<p>We highly recommend using the <a href="https://www.ebi.ac.uk/ols4/">EMBL-EBI Ontology Lookup Service</a>
to find valid IDs for your metadata.</p>
<p>Models have a limited acceptable range of metadata input values. If
you provide a value that is not in the acceptable range, the API will
return an error.</p>
</div>
<div id="query-parameters" class="section level3">
<h3>Query Parameters</h3>
<p>In addition to metadata, queries support several optional parameters
that control the generation process:</p>
<div id="mode-character-required" class="section level4">
<h4>mode (character, required)</h4>
<p>Controls the type of prediction the model generates. This parameter
is required in all queries.</p>
<p>Available modes:</p>
<ul>
<li><p><strong>“sample generation”</strong>: The model works identically
to the mean estimation approach, except that the final gene expression
distribution is also sampled to generate realistic-looking synthetic
data that captures the error associated with measurements. This mode is
useful when you want data that mimics real experimental
measurements.</p></li>
<li><p><strong>“mean estimation”</strong>: The model creates a
distribution capturing the biological heterogeneity consistent with the
supplied metadata. This distribution is then sampled to predict a gene
expression distribution that captures measurement error. The mean of
that distribution serves as the prediction. This mode is useful when you
want a stable estimate of expected expression levels.</p></li>
</ul>
<blockquote>
<p><strong>Note:</strong> <strong>Single-cell queries only support “mean
estimation” mode.</strong> Bulk queries support both modes.</p>
</blockquote>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a><span class="co"># Bulk query with sample generation (default for bulk)</span></span>
<span id="cb12-2"><a href="#cb12-2" tabindex="-1"></a>bulk_query <span class="ot">&lt;-</span> <span class="fu">get_valid_query</span>(<span class="at">modality =</span> <span class="st">&quot;bulk&quot;</span>)</span>
<span id="cb12-3"><a href="#cb12-3" tabindex="-1"></a>bulk_query<span class="sc">$</span>mode <span class="ot">&lt;-</span> <span class="st">&quot;sample generation&quot;</span></span>
<span id="cb12-4"><a href="#cb12-4" tabindex="-1"></a></span>
<span id="cb12-5"><a href="#cb12-5" tabindex="-1"></a><span class="co"># Bulk query with mean estimation</span></span>
<span id="cb12-6"><a href="#cb12-6" tabindex="-1"></a>bulk_query_mean <span class="ot">&lt;-</span> <span class="fu">get_valid_query</span>(<span class="at">modality =</span> <span class="st">&quot;bulk&quot;</span>)</span>
<span id="cb12-7"><a href="#cb12-7" tabindex="-1"></a>bulk_query_mean<span class="sc">$</span>mode <span class="ot">&lt;-</span> <span class="st">&quot;mean estimation&quot;</span></span>
<span id="cb12-8"><a href="#cb12-8" tabindex="-1"></a></span>
<span id="cb12-9"><a href="#cb12-9" tabindex="-1"></a><span class="co"># Single-cell query (must use mean estimation)</span></span>
<span id="cb12-10"><a href="#cb12-10" tabindex="-1"></a>sc_query <span class="ot">&lt;-</span> <span class="fu">get_valid_query</span>(<span class="at">modality =</span> <span class="st">&quot;single-cell&quot;</span>)</span>
<span id="cb12-11"><a href="#cb12-11" tabindex="-1"></a>sc_query<span class="sc">$</span>mode <span class="ot">&lt;-</span> <span class="st">&quot;mean estimation&quot;</span> <span class="co"># Required for single-cell</span></span></code></pre></div>
</div>
<div id="total_count-integer-optional" class="section level4">
<h4>total_count (integer, optional)</h4>
<p>Library size used when converting predicted log CPM back to raw
counts. Higher values scale counts up proportionally.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a><span class="co"># Create a query and add custom total_count</span></span>
<span id="cb13-2"><a href="#cb13-2" tabindex="-1"></a>query <span class="ot">&lt;-</span> <span class="fu">get_valid_query</span>(<span class="at">modality =</span> <span class="st">&quot;bulk&quot;</span>)</span>
<span id="cb13-3"><a href="#cb13-3" tabindex="-1"></a>query<span class="sc">$</span>total_count <span class="ot">&lt;-</span> <span class="dv">5000000</span></span></code></pre></div>
</div>
<div id="deterministic_latents-logical-optional" class="section level4">
<h4>deterministic_latents (logical, optional)</h4>
<p>If <code>TRUE</code>, the model uses the mean of each latent
distribution (<code>p(z|metadata)</code> or <code>q(z|x)</code>) instead
of sampling. This removes randomness from latent sampling and produces
deterministic outputs for the same inputs.</p>
<ul>
<li>Default: <code>FALSE</code> (sampling is enabled)</li>
</ul>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a><span class="co"># Create a query and enable deterministic latents</span></span>
<span id="cb14-2"><a href="#cb14-2" tabindex="-1"></a>query <span class="ot">&lt;-</span> <span class="fu">get_valid_query</span>(<span class="at">modality =</span> <span class="st">&quot;bulk&quot;</span>)</span>
<span id="cb14-3"><a href="#cb14-3" tabindex="-1"></a>query<span class="sc">$</span>deterministic_latents <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span></code></pre></div>
</div>
<div id="seed-integer-optional" class="section level4">
<h4>seed (integer, optional)</h4>
<p>Random seed for reproducibility when using stochastic sampling.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a><span class="co"># Create a query with a specific seed</span></span>
<span id="cb15-2"><a href="#cb15-2" tabindex="-1"></a>query <span class="ot">&lt;-</span> <span class="fu">get_valid_query</span>(<span class="at">modality =</span> <span class="st">&quot;bulk&quot;</span>)</span>
<span id="cb15-3"><a href="#cb15-3" tabindex="-1"></a>query<span class="sc">$</span>seed <span class="ot">&lt;-</span> <span class="dv">42</span></span></code></pre></div>
<p>You can combine multiple parameters in a single query:</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" tabindex="-1"></a><span class="co"># Create a query and add multiple parameters</span></span>
<span id="cb16-2"><a href="#cb16-2" tabindex="-1"></a>query <span class="ot">&lt;-</span> <span class="fu">get_valid_query</span>(<span class="at">modality =</span> <span class="st">&quot;bulk&quot;</span>)</span>
<span id="cb16-3"><a href="#cb16-3" tabindex="-1"></a>query<span class="sc">$</span>total_count <span class="ot">&lt;-</span> <span class="dv">8000000</span></span>
<span id="cb16-4"><a href="#cb16-4" tabindex="-1"></a>query<span class="sc">$</span>deterministic_latents <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb16-5"><a href="#cb16-5" tabindex="-1"></a>query<span class="sc">$</span>mode <span class="ot">&lt;-</span> <span class="st">&quot;mean estimation&quot;</span></span>
<span id="cb16-6"><a href="#cb16-6" tabindex="-1"></a></span>
<span id="cb16-7"><a href="#cb16-7" tabindex="-1"></a>results <span class="ot">&lt;-</span> <span class="fu">predict_query</span>(query)</span></code></pre></div>
</div>
</div>
<div id="modifying-query-inputs" class="section level3">
<h3>Modifying Query Inputs</h3>
<p>You can customize the query inputs to fit your specific research
needs:</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" tabindex="-1"></a><span class="co"># Get a base query</span></span>
<span id="cb17-2"><a href="#cb17-2" tabindex="-1"></a>query <span class="ot">&lt;-</span> <span class="fu">get_valid_query</span>()</span>
<span id="cb17-3"><a href="#cb17-3" tabindex="-1"></a></span>
<span id="cb17-4"><a href="#cb17-4" tabindex="-1"></a><span class="co"># Adjust number of samples for the first input</span></span>
<span id="cb17-5"><a href="#cb17-5" tabindex="-1"></a>query<span class="sc">$</span>inputs[[<span class="dv">1</span>]]<span class="sc">$</span>num_samples <span class="ot">&lt;-</span> <span class="dv">10</span></span>
<span id="cb17-6"><a href="#cb17-6" tabindex="-1"></a></span>
<span id="cb17-7"><a href="#cb17-7" tabindex="-1"></a><span class="co"># Add a new condition</span></span>
<span id="cb17-8"><a href="#cb17-8" tabindex="-1"></a>query<span class="sc">$</span>inputs[[<span class="dv">3</span>]] <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb17-9"><a href="#cb17-9" tabindex="-1"></a>  <span class="at">metadata =</span> <span class="fu">list</span>(</span>
<span id="cb17-10"><a href="#cb17-10" tabindex="-1"></a>    <span class="at">sex =</span> <span class="st">&quot;male&quot;</span>,</span>
<span id="cb17-11"><a href="#cb17-11" tabindex="-1"></a>    <span class="at">sample_type =</span> <span class="st">&quot;primary tissue&quot;</span>,</span>
<span id="cb17-12"><a href="#cb17-12" tabindex="-1"></a>    <span class="at">tissue_ontology_id =</span> <span class="st">&quot;UBERON:0002371&quot;</span></span>
<span id="cb17-13"><a href="#cb17-13" tabindex="-1"></a>  ),</span>
<span id="cb17-14"><a href="#cb17-14" tabindex="-1"></a>  <span class="at">num_samples =</span> <span class="dv">5</span></span>
<span id="cb17-15"><a href="#cb17-15" tabindex="-1"></a>)</span></code></pre></div>
</div>
<div id="working-with-results" class="section level3">
<h3>Working with Results</h3>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" tabindex="-1"></a><span class="co"># Access metadata and expression matrices</span></span>
<span id="cb18-2"><a href="#cb18-2" tabindex="-1"></a>metadata <span class="ot">&lt;-</span> result<span class="sc">$</span>metadata</span>
<span id="cb18-3"><a href="#cb18-3" tabindex="-1"></a>expression <span class="ot">&lt;-</span> result<span class="sc">$</span>expression</span>
<span id="cb18-4"><a href="#cb18-4" tabindex="-1"></a></span>
<span id="cb18-5"><a href="#cb18-5" tabindex="-1"></a><span class="co"># Check dimensions</span></span>
<span id="cb18-6"><a href="#cb18-6" tabindex="-1"></a><span class="fu">dim</span>(expression)</span>
<span id="cb18-7"><a href="#cb18-7" tabindex="-1"></a></span>
<span id="cb18-8"><a href="#cb18-8" tabindex="-1"></a><span class="co"># View metadata sample</span></span>
<span id="cb18-9"><a href="#cb18-9" tabindex="-1"></a><span class="fu">head</span>(metadata)</span></code></pre></div>
<p>You may want to process the data in chunks or save it for later
use:</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" tabindex="-1"></a><span class="co"># Save results to RDS file</span></span>
<span id="cb19-2"><a href="#cb19-2" tabindex="-1"></a><span class="fu">saveRDS</span>(result, <span class="st">&quot;synthesize_results.rds&quot;</span>)</span>
<span id="cb19-3"><a href="#cb19-3" tabindex="-1"></a></span>
<span id="cb19-4"><a href="#cb19-4" tabindex="-1"></a><span class="co"># Load previously saved results</span></span>
<span id="cb19-5"><a href="#cb19-5" tabindex="-1"></a>result <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">&quot;synthesize_results.rds&quot;</span>)</span>
<span id="cb19-6"><a href="#cb19-6" tabindex="-1"></a></span>
<span id="cb19-7"><a href="#cb19-7" tabindex="-1"></a><span class="co"># Export as CSV</span></span>
<span id="cb19-8"><a href="#cb19-8" tabindex="-1"></a><span class="fu">write.csv</span>(result<span class="sc">$</span>expression, <span class="st">&quot;expression_matrix.csv&quot;</span>)</span>
<span id="cb19-9"><a href="#cb19-9" tabindex="-1"></a><span class="fu">write.csv</span>(result<span class="sc">$</span>metadata, <span class="st">&quot;sample_metadata.csv&quot;</span>)</span></code></pre></div>
</div>
<div id="custom-validation" class="section level3">
<h3>Custom Validation</h3>
<p>You can validate your queries before sending them to the API:</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" tabindex="-1"></a><span class="co"># Validate structure</span></span>
<span id="cb20-2"><a href="#cb20-2" tabindex="-1"></a><span class="fu">validate_query</span>(query)</span>
<span id="cb20-3"><a href="#cb20-3" tabindex="-1"></a></span>
<span id="cb20-4"><a href="#cb20-4" tabindex="-1"></a><span class="co"># Validate modality</span></span>
<span id="cb20-5"><a href="#cb20-5" tabindex="-1"></a><span class="fu">validate_modality</span>(query)</span></code></pre></div>
</div>
</div>
<div id="session-info" class="section level2">
<h2>Session info</h2>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code></pre></div>
</div>
<div id="additional-resources" class="section level2">
<h2>Additional Resources</h2>
<ul>
<li><a href="https://github.com/synthesizebio/rsynthbio">Package Source
Code</a></li>
<li><a href="https://github.com/synthesizebio/rsynthbio/issues">File Bug
Reports</a></li>
</ul>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
