<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />



<title>Reference Conditioning</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Reference Conditioning</h1>



<div id="overview" class="section level2">
<h2>Overview</h2>
<p>Reference conditioning models generate expression data
<strong>conditioned on a real reference sample</strong>. This allows you
to “anchor” to an existing expression profile while applying
perturbations or modifications.</p>
<p>This is useful when you want to:</p>
<ul>
<li>Simulate the effect of a perturbation on a specific sample</li>
<li>Generate expression profiles that preserve the biological and
technical characteristics of a reference</li>
<li>Create synthetic “treated vs. control” pairs</li>
</ul>
</div>
<div id="available-models" class="section level2">
<h2>Available Models</h2>
<ul>
<li><strong><code>gem-1-bulk_reference-conditioning</code></strong>:
Bulk RNA-seq reference conditioning model</li>
<li><strong><code>gem-1-sc_reference-conditioning</code></strong>:
Single-cell RNA-seq reference conditioning model</li>
</ul>
<blockquote>
<p><strong>Note:</strong> These endpoints may require 1-2 minutes of
startup time if they have been scaled down. Plan accordingly for
interactive use.</p>
</blockquote>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">library</span>(rsynthbio)</span></code></pre></div>
</div>
<div id="how-it-works" class="section level2">
<h2>How It Works</h2>
<p>Reference conditioning encodes the biological and technical
characteristics from a real expression sample, then generates new
expression data that:</p>
<ol style="list-style-type: decimal">
<li>Preserves the biological/technical latent space of the
reference</li>
<li>Applies any perturbation metadata you specify</li>
<li>Returns synthetic expression that reflects the perturbation effect
on that specific sample</li>
</ol>
</div>
<div id="creating-a-query" class="section level2">
<h2>Creating a Query</h2>
<p>Reference conditioning queries require different inputs than baseline
models:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="co"># Get the example query structure</span></span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a>example_query <span class="ot">&lt;-</span> <span class="fu">get_example_query</span>(<span class="at">model_id =</span> <span class="st">&quot;gem-1-bulk_reference-conditioning&quot;</span>)<span class="sc">$</span>example_query</span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a></span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a><span class="co"># Inspect the query structure</span></span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a><span class="fu">str</span>(example_query)</span></code></pre></div>
<p>The query structure includes:</p>
<ol style="list-style-type: decimal">
<li><p><strong><code>inputs</code></strong>: A list where each input
contains:</p>
<ul>
<li><strong><code>counts</code></strong>: The reference expression
counts (a named list with a <code>counts</code> vector)</li>
<li><strong><code>metadata</code></strong>: Perturbation-only metadata
(see below)</li>
<li><strong><code>num_samples</code></strong>: How many samples to
generate</li>
</ul></li>
<li><p><strong><code>conditioning</code></strong>: Which latent spaces
to condition on (typically
<code>[&quot;biological&quot;, &quot;technical&quot;]</code>)</p></li>
<li><p><strong><code>sampling_strategy</code></strong>:
<code>&quot;mean estimation&quot;</code> or
<code>&quot;sample generation&quot;</code></p></li>
</ol>
<div id="perturbation-only-metadata" class="section level3">
<h3>Perturbation-Only Metadata</h3>
<p>Unlike baseline models, reference conditioning queries only accept
perturbation metadata fields:</p>
<ul>
<li><code>perturbation_ontology_id</code></li>
<li><code>perturbation_type</code></li>
<li><code>perturbation_time</code></li>
<li><code>perturbation_dose</code></li>
</ul>
<p>All other biological and technical metadata is inferred from the
reference expression.</p>
</div>
</div>
<div id="example-simulating-a-drug-treatment" class="section level2">
<h2>Example: Simulating a Drug Treatment</h2>
<p>Here’s a complete example simulating a drug treatment effect on a
reference sample:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="co"># Start with example query structure</span></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a>query <span class="ot">&lt;-</span> <span class="fu">get_example_query</span>(<span class="at">model_id =</span> <span class="st">&quot;gem-1-bulk_reference-conditioning&quot;</span>)<span class="sc">$</span>example_query</span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a></span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a><span class="co"># Replace with your actual reference counts</span></span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a><span class="co"># The counts vector must match the model&#39;s expected gene order and length</span></span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a>query<span class="sc">$</span>inputs[[<span class="dv">1</span>]]<span class="sc">$</span>counts <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">counts =</span> your_reference_counts)</span>
<span id="cb3-7"><a href="#cb3-7" tabindex="-1"></a></span>
<span id="cb3-8"><a href="#cb3-8" tabindex="-1"></a><span class="co"># Specify the perturbation</span></span>
<span id="cb3-9"><a href="#cb3-9" tabindex="-1"></a>query<span class="sc">$</span>inputs[[<span class="dv">1</span>]]<span class="sc">$</span>metadata <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb3-10"><a href="#cb3-10" tabindex="-1"></a>  <span class="at">perturbation_ontology_id =</span> <span class="st">&quot;CHEMBL25&quot;</span>, <span class="co"># Aspirin (ChEMBL ID)</span></span>
<span id="cb3-11"><a href="#cb3-11" tabindex="-1"></a>  <span class="at">perturbation_type =</span> <span class="st">&quot;compound&quot;</span>,</span>
<span id="cb3-12"><a href="#cb3-12" tabindex="-1"></a>  <span class="at">perturbation_time =</span> <span class="st">&quot;24h&quot;</span>,</span>
<span id="cb3-13"><a href="#cb3-13" tabindex="-1"></a>  <span class="at">perturbation_dose =</span> <span class="st">&quot;10uM&quot;</span></span>
<span id="cb3-14"><a href="#cb3-14" tabindex="-1"></a>)</span>
<span id="cb3-15"><a href="#cb3-15" tabindex="-1"></a></span>
<span id="cb3-16"><a href="#cb3-16" tabindex="-1"></a>query<span class="sc">$</span>inputs[[<span class="dv">1</span>]]<span class="sc">$</span>num_samples <span class="ot">&lt;-</span> <span class="dv">3</span></span>
<span id="cb3-17"><a href="#cb3-17" tabindex="-1"></a></span>
<span id="cb3-18"><a href="#cb3-18" tabindex="-1"></a><span class="co"># Set the sampling strategy</span></span>
<span id="cb3-19"><a href="#cb3-19" tabindex="-1"></a>query<span class="sc">$</span>sampling_strategy <span class="ot">&lt;-</span> <span class="st">&quot;mean estimation&quot;</span></span>
<span id="cb3-20"><a href="#cb3-20" tabindex="-1"></a></span>
<span id="cb3-21"><a href="#cb3-21" tabindex="-1"></a><span class="co"># Submit the query</span></span>
<span id="cb3-22"><a href="#cb3-22" tabindex="-1"></a>result <span class="ot">&lt;-</span> <span class="fu">predict_query</span>(query, <span class="at">model_id =</span> <span class="st">&quot;gem-1-bulk_reference-conditioning&quot;</span>)</span></code></pre></div>
</div>
<div id="example-crispr-knockout-simulation" class="section level2">
<h2>Example: CRISPR Knockout Simulation</h2>
<p>Simulate the effect of knocking out a specific gene:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a>query <span class="ot">&lt;-</span> <span class="fu">get_example_query</span>(<span class="at">model_id =</span> <span class="st">&quot;gem-1-bulk_reference-conditioning&quot;</span>)<span class="sc">$</span>example_query</span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a></span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a><span class="co"># Your reference sample counts</span></span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a>query<span class="sc">$</span>inputs[[<span class="dv">1</span>]]<span class="sc">$</span>counts <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">counts =</span> control_sample_counts)</span>
<span id="cb4-5"><a href="#cb4-5" tabindex="-1"></a></span>
<span id="cb4-6"><a href="#cb4-6" tabindex="-1"></a><span class="co"># CRISPR knockout of TP53</span></span>
<span id="cb4-7"><a href="#cb4-7" tabindex="-1"></a>query<span class="sc">$</span>inputs[[<span class="dv">1</span>]]<span class="sc">$</span>metadata <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb4-8"><a href="#cb4-8" tabindex="-1"></a>  <span class="at">perturbation_ontology_id =</span> <span class="st">&quot;ENSG00000141510&quot;</span>, <span class="co"># TP53 Ensembl ID</span></span>
<span id="cb4-9"><a href="#cb4-9" tabindex="-1"></a>  <span class="at">perturbation_type =</span> <span class="st">&quot;crispr&quot;</span></span>
<span id="cb4-10"><a href="#cb4-10" tabindex="-1"></a>)</span>
<span id="cb4-11"><a href="#cb4-11" tabindex="-1"></a></span>
<span id="cb4-12"><a href="#cb4-12" tabindex="-1"></a>query<span class="sc">$</span>inputs[[<span class="dv">1</span>]]<span class="sc">$</span>num_samples <span class="ot">&lt;-</span> <span class="dv">5</span></span>
<span id="cb4-13"><a href="#cb4-13" tabindex="-1"></a></span>
<span id="cb4-14"><a href="#cb4-14" tabindex="-1"></a>result <span class="ot">&lt;-</span> <span class="fu">predict_query</span>(query, <span class="at">model_id =</span> <span class="st">&quot;gem-1-bulk_reference-conditioning&quot;</span>)</span></code></pre></div>
</div>
<div id="query-parameters" class="section level2">
<h2>Query Parameters</h2>
<div id="conditioning-list-optional" class="section level3">
<h3>conditioning (list, optional)</h3>
<p>Controls which latent spaces are conditioned on the reference.
Default is <code>[&quot;biological&quot;, &quot;technical&quot;]</code>.</p>
<p>When both are conditioned, the model preserves both biological
identity and technical characteristics from the reference sample.</p>
</div>
<div id="sampling_strategy-character-required" class="section level3">
<h3>sampling_strategy (character, required)</h3>
<p>Controls the type of prediction:</p>
<ul>
<li><strong>“sample generation”</strong>: Generates realistic-looking
synthetic data with measurement error. <strong>(Bulk only)</strong></li>
<li><strong>“mean estimation”</strong>: Provides stable mean estimates.
<strong>(Bulk and single-cell)</strong></li>
</ul>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a>query<span class="sc">$</span>sampling_strategy <span class="ot">&lt;-</span> <span class="st">&quot;mean estimation&quot;</span></span></code></pre></div>
</div>
<div id="fixed_total_count-logical-optional" class="section level3">
<h3>fixed_total_count (logical, optional)</h3>
<p>Controls whether to preserve the reference’s library size:</p>
<ul>
<li><strong><code>FALSE</code></strong> (default): The output’s total
count is taken from the reference expression (sum of its counts). Use
this when you want the synthetic sample to preserve the reference’s
library size.</li>
<li><strong><code>TRUE</code></strong>: Forces the model to use the
<code>total_count</code> parameter value (or default) instead of the
reference’s library size.</li>
</ul>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="co"># Preserve reference library size (default)</span></span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a>query<span class="sc">$</span>fixed_total_count <span class="ot">&lt;-</span> <span class="cn">FALSE</span></span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a></span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a><span class="co"># Or force a specific library size</span></span>
<span id="cb6-5"><a href="#cb6-5" tabindex="-1"></a>query<span class="sc">$</span>fixed_total_count <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb6-6"><a href="#cb6-6" tabindex="-1"></a>query<span class="sc">$</span>total_count <span class="ot">&lt;-</span> <span class="dv">10000000</span></span></code></pre></div>
</div>
<div id="total_count-integer-optional" class="section level3">
<h3>total_count (integer, optional)</h3>
<p>Library size used when converting predicted log CPM back to raw
counts. Only effective when <code>fixed_total_count = TRUE</code>.</p>
<ul>
<li>Default: 10,000,000 for bulk; 10,000 for single-cell</li>
</ul>
</div>
<div id="deterministic_latents-logical-optional" class="section level3">
<h3>deterministic_latents (logical, optional)</h3>
<p>If <code>TRUE</code>, the model uses the mean of each latent
distribution (<code>p(z|metadata)</code> for perturbation,
<code>q(z|x)</code> for conditioned components) instead of sampling.
This produces deterministic, reproducible outputs.</p>
<ul>
<li>Default: <code>FALSE</code></li>
</ul>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a>query<span class="sc">$</span>deterministic_latents <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span></code></pre></div>
</div>
<div id="seed-integer-optional" class="section level3">
<h3>seed (integer, optional)</h3>
<p>Random seed for reproducibility.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a>query<span class="sc">$</span>seed <span class="ot">&lt;-</span> <span class="dv">42</span></span></code></pre></div>
</div>
</div>
<div id="valid-perturbation-metadata" class="section level2">
<h2>Valid Perturbation Metadata</h2>
<table>
<colgroup>
<col width="13%" />
<col width="86%" />
</colgroup>
<thead>
<tr class="header">
<th>Field</th>
<th>Description / Format</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>perturbation_ontology_id</code></td>
<td>Ensembl gene ID (e.g., <code>ENSG00000141510</code>), <a href="https://www.ebi.ac.uk/chebi/">ChEBI ID</a>, <a href="https://www.ebi.ac.uk/chembl/">ChEMBL ID</a>, or <a href="https://www.ncbi.nlm.nih.gov/taxonomy">NCBI Taxonomy ID</a></td>
</tr>
<tr class="even">
<td><code>perturbation_type</code></td>
<td>One of: “coculture”, “compound”, “control”, “crispr”, “genetic”,
“infection”, “other”, “overexpression”, “peptide or biologic”, “shrna”,
“sirna”</td>
</tr>
<tr class="odd">
<td><code>perturbation_time</code></td>
<td>Time since perturbation (e.g., “24h”, “48h”)</td>
</tr>
<tr class="even">
<td><code>perturbation_dose</code></td>
<td>Dose of perturbation (e.g., “10uM”, “1mg/kg”)</td>
</tr>
</tbody>
</table>
</div>
<div id="working-with-results" class="section level2">
<h2>Working with Results</h2>
<p>The result structure is similar to baseline models:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="co"># Access metadata and expression matrices</span></span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a>metadata <span class="ot">&lt;-</span> result<span class="sc">$</span>metadata</span>
<span id="cb9-3"><a href="#cb9-3" tabindex="-1"></a>expression <span class="ot">&lt;-</span> result<span class="sc">$</span>expression</span>
<span id="cb9-4"><a href="#cb9-4" tabindex="-1"></a></span>
<span id="cb9-5"><a href="#cb9-5" tabindex="-1"></a><span class="co"># Compare to your reference</span></span>
<span id="cb9-6"><a href="#cb9-6" tabindex="-1"></a><span class="fu">dim</span>(expression)</span>
<span id="cb9-7"><a href="#cb9-7" tabindex="-1"></a><span class="fu">head</span>(metadata)</span></code></pre></div>
<div id="differential-expression" class="section level3">
<h3>Differential Expression</h3>
<p>When conditioning on both biological and technical latents, you can
directly compare the generated expression to your reference to identify
perturbation effects:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="co"># Your reference (input) counts</span></span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a>reference_cpm <span class="ot">&lt;-</span> your_reference_counts <span class="sc">/</span> <span class="fu">sum</span>(your_reference_counts) <span class="sc">*</span> <span class="fl">1e6</span></span>
<span id="cb10-3"><a href="#cb10-3" tabindex="-1"></a></span>
<span id="cb10-4"><a href="#cb10-4" tabindex="-1"></a><span class="co"># Generated (perturbed) counts</span></span>
<span id="cb10-5"><a href="#cb10-5" tabindex="-1"></a>generated_cpm <span class="ot">&lt;-</span> expression[<span class="dv">1</span>, ] <span class="sc">/</span> <span class="fu">sum</span>(expression[<span class="dv">1</span>, ]) <span class="sc">*</span> <span class="fl">1e6</span></span>
<span id="cb10-6"><a href="#cb10-6" tabindex="-1"></a></span>
<span id="cb10-7"><a href="#cb10-7" tabindex="-1"></a><span class="co"># Log fold change</span></span>
<span id="cb10-8"><a href="#cb10-8" tabindex="-1"></a>log2fc <span class="ot">&lt;-</span> <span class="fu">log2</span>(generated_cpm <span class="sc">+</span> <span class="dv">1</span>) <span class="sc">-</span> <span class="fu">log2</span>(reference_cpm <span class="sc">+</span> <span class="dv">1</span>)</span>
<span id="cb10-9"><a href="#cb10-9" tabindex="-1"></a></span>
<span id="cb10-10"><a href="#cb10-10" tabindex="-1"></a><span class="co"># Identify top changed genes</span></span>
<span id="cb10-11"><a href="#cb10-11" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">sort</span>(log2fc, <span class="at">decreasing =</span> <span class="cn">TRUE</span>), <span class="dv">20</span>)</span></code></pre></div>
</div>
</div>
<div id="important-notes" class="section level2">
<h2>Important Notes</h2>
<div id="counts-vector-length" class="section level3">
<h3>Counts Vector Length</h3>
<p>The reference counts vector must match the model’s expected number of
genes. If the length doesn’t match, the API will return a validation
error.</p>
<p>Use <code>get_example_query()</code> to see the expected structure
and ensure your counts vector has the correct length.</p>
</div>
<div id="gene-order" class="section level3">
<h3>Gene Order</h3>
<p>Ensure your reference counts are in the same gene order expected by
the model. The response includes a <code>gene_order</code> field that
specifies the expected order.</p>
</div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
